### ISSUE TITLE
Enforce offline availability of external dependencies before execution

### ISSUE THREAD

**Issue by tonit** [gradle/gradle#5950]
I want to proactively cache all external resources (dependencies, plugin dependencies or anything that Gradle would download & cache normally) without requiring to run the actual build.

The goal is to create a bag of "this is everything you will need if actually run" and later use it in subsequent builds.
The only way to achieve this at the moment reliably is what i call "mirror builds".
This means, running the build once which fills the caches and then i can safely run with "--offline" later.

The problems with that are:
1. Unnecessary work. For example in my case, a full build on a monorepo takes 30 mins on a 4-core machine. Most time is spent running tests. So, if that build only should cache the dependencies, you don't actually need to execute the tests. Note that tests are only one example and here junit-engine is the artefact that only gets requested (thus cached) when the test task gets executed.
1. The other issue with that is the cache is not portable really. It is a cache, right? So yes i can copy caches from one machine to another, I treat them as a blackbox. So there is no easy way to (safely) populate 100% of the cache to a nexus repo or any other m2 repo, correct?

To summarise, i want to create a bucket for a given gradle execution, say: "gradle build" to spit out a m2 repo. (remember the issues mentioned above). Then, when run "gradle build" with the same sources on a different machine with only that m2 repo, it all should work.

As a comparison, Bazelbuild is doing something like this i was told.

I am somewhat pessimistic if something like this is possible without breaking existing plugins. But i am not a gradle engineer, so maybe there is a way.

**Comment by oehme** [comment]
There is a [community plugin](https://github.com/mdietrichstein/gradle-offline-dependencies-plugin) for this. We currently don't have plans to provide this functionality inside Gradle's core.

**Comment by tonit** [comment]
A plugin won't cut it as the issue is core to gradle (as outlined above). However you just said it: "don't have plans to provide this functionality inside Gradle's core".
Guessed it. Just followed a suggestion from @melix https://twitter.com/CedricChampeau/status/1017332044047798273

**Comment by oehme** [comment]
I don't understand what you mean by "a plugin won't cut it" - The plugin I linked does what you are asking for. It mirrors your dependencies so you can then use them as a Maven repo.

**Comment by tonit** [comment]
Great, @melix & @oehme. Will use your suggestions and see if it fits my need. Thanks!

**Comment by ThadHouse** [comment]
With the changes made to offline caching in 4.10, I definitely see a much greater reason to incorporate this into gradle core. The existing offline plugin uses some internal API's, and actually broke in gradle 4.9. So either maybe you guys could help find public API replacements for their internal APIs, add public APIs if none exist, or add a way into gradle core to build a proper offline cache.
